sycl/doc/extensions/proposed/sycl_ext_oneapi_device_image_backend_content.asciidoc

```diff
@@ -0,0 +1,234 @@
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 9 specification.
+All references below to the "core SYCL specification" or to section numbers in
+the SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.
+Interfaces defined in this specification may not be implemented yet or may be
+in a preliminary state.
+The specification itself may also change in incompatible ways before it is
+finalized.
+*Shipping software products should not rely on APIs defined in this
+specification.*
+
+
+== Overview
+
+This extension adds a mechanism to obtain the raw backend content of the device
+images that are in a kernel bundle.
+The format of this content is implementation-defined, so applications that make
+use of this extension are normally not portable to other implementations of
+SYCL.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_DEVICE_IMAGE_BACKEND_CONTENT` to one of the values defined in
+the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New member functions in the `device_image` class
+
+This extension adds the following member functions to the `device_image` class.
+
+[source,c++]
+----
+namespace sycl {
+
+template <bundle_state State>
+class device_image {
+ public:
+  std::vector<std::byte> ext_oneapi_get_backend_content() const;
+  std::span<std::byte> ext_oneapi_get_backend_content_view() const;
+
+  /*...*/
+};
+
+} // namespace sycl
+----
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+std::vector<std::byte> ext_oneapi_get_backend_content() const;
+----
+!====
+
+_Constraints:_ Available only when `State` is `bundle_state::executable`.
+
+_Returns:_ A copy of the raw backend content for this device image.
+The format of this data is implementation-defined.
+See below for a description of this format on {dpcpp}.
+
+'''
+
+[frame=all,grid=none,separator="@"]
+!====
+a@
+[source,c++]
+----
+std::span<std::byte> ext_oneapi_get_content_backend_view() const;
+----
+!====
+
+Available only when the compiler is {cpp}20 or higher.
```

Comment:
I think we should drop this, and rely on the `ifdef` in the synopsis.  This is really saying "Available only when the library provides an implementation of `std::span`", and that's implied by usage of the type.

Reply:
Let's try to agree on standard wording for cases like this because I think it will come up more and more.

I feel like we should say somewhere that this API is tied to C++20.  How do you feel about the wording I added in d288119e81034f06752cc00187bd43b042abe015?

Reply:
I would have said this is good enough, but if you'd like to settle on standard wording I'd like to talk about it a bit more.

I think it would be more accurate to say that the _implementation_ defines the macro.  [ISO C++](https://eel.is/c++draft/version.syn#:macro,feature-test) says the implementation only defines the macro if you include `<version>` or `<span>`.

Given the above, do you want to say anything about who is responsible for including `<span>` somewhere?  It probably shouldn't be part of the function synopsis, but we should make it clear that if this extension is implemented then `sycl.hpp` is expected to include `<span>` if it's available (see [`__has_include`](https://eel.is/c++draft/cpp.cond#15)), and that this isn't a user's responsibility. For the actual SYCL specification, I imagine we'd have a single dedicated section to explain how compilation with different versions of C++ works, and we could cover this there instead of in each extension.

Finally, I think we could improve the formatting.  We should add a title to the paragraph (like we do with _Constraints_, etc) and consider rephrasing the "Available only when" (to further differentiate from constraints).  How about something like:

> _Required C++ Features_: The implementation must define the `__cpp_lib_span` feature-test macro (which is defined in C++20 and higher).

I deliberately put "C++" in the title in an attempt to be future-proof.  Even if ISO C++ decided to introduce a similar concept with the same name, it would probably just be called "Required Features" (since the C++ would be implied).

Reply:
It looks like this is going to be a longer conversation, and I'd like to get implementation started on this extension.  Would you be OK approving what we have now, and I'll open an issue in the Khronos repo to continue discussion about the precise wording that we'll use going forward?

Reply:
Yes, absolutely!

Reply:
Issue opened: KhronosGroup/SYCL-Docs#667